### ISSUE TITLE
mrcalc: parses nonexistent filenames that start with numbers as numbers

### ISSUE THREAD

**Issue by PatronBernard** [MRtrix3/mrtrix3#1780]
**Bug description**

`mrcalc` treats nonexisting files with numbers in as valid operands.

**To Reproduce**

```
mrcalc 314IAmNotARealFile.mif 2 -pow
#Outputs 98596
```
Similarly:
```
mrcalc 314IAmNotARealFilemif 2 -pow
#Outputs 98596
```

**Platform/Environment/Version**
-  Ubuntu 16.04/
-  *MRtrix3* version: == mrinfo 3.0_RC3-1083-

Greetings,
Jan Morez
Visionlab, UAntwerpen

**Comment by Lestropie** [comment]
Cheers for the report Jan!

If such a file actually existed, then the [relevant `mrcalc` code](https://github.com/MRtrix3/mrtrix3/blob/3.0_RC3/cmd/mrcalc.cpp#L238-L258) should load that string as an image before it attempts to parse the string as a numerical value; so the bug is indeed specific to a non-existing file.

I actually encountered the fundamental problem underlying this issue recently in #1738. The relevant code commences [here](https://github.com/MRtrix3/mrtrix3/pull/1738/files#diff-197414e741ab8ae2da0a4778de50766eR144). The problem goes *all* the way down to the `MR::to<>()` function, which converts a string to some other type using `istringstream`'s `>>` operator. It is possible for the *start* of a string to be successfully converted to some other type, but for the remainder of the string to be *ignored*, and the `istringstream` will indicate that the conversion was successful. This is why the string "`314IAmNotARealFile.mif`" happily converts to the integer 314. What I do in #1738 is attempt such a conversion, but only accept the conversion *if the entire string is utilised* as part of that conversion.

So I believe the solution to this issue is going to be to take that concept, and implement it in `MR::to<>()`.

**Comment by jdtournier** [comment]
> So I believe the solution to this issue is going to be to take that concept, and implement it in MR::to<>().

Yep, happy with that. :+1:

**Comment by Lestropie** [comment]
Didn't finalise a solution to this yesterday... Encountered this oddity before I left:

```
std::stringstream s ("1.mif");
default_type temp;
s >> temp;
std::cerr << s.eof();
```
False

```
std::stringstream s ("1.mif");
std::complex<T> temp;
s >> temp;
std::cerr << s.eof();
```
True

:face_with_head_bandage:

**Comment by jdtournier** [comment]
I think this is fixed now (at least on `dev`) with #1794 - closing.
